### ISSUE TITLE
Accessing registered variables created from commands with braces results in template processing errors

### ISSUE THREAD

**Issue by enxebre** [ansible/ansible-modules-core#1641]
Hi,
running from os x mavericks 10.9 to ubuntu1404
ansible 1.9.0.1
docker Client version: 1.7.0 Client API version: 1.19
ansible.cfg

```
[defaults]
host_key_checking = no
record_host_keys = no
jinja2_extensions = jinja2.ext.do
timeout = 15
gathering = smart

[ssh_connection]
ssh_args = -o ControlMaster=auto -o ControlPersist=30m
scp_if_ssh = True
pipelining = True
```

trying to:

```
- name: ensure cadvisor is running (and enable it at boot)
  when: cadvisor_enabled
  sudo: yes
  service:
    name: cadvisor
    state: started
    enabled: yes
  tags:
    - cadvisor

- name: cadvisor container ip
  sudo: yes
  command: "{% raw %} docker inspect --format '{{ .NetworkSettings.IPAddress }}' cadvisor {% endraw %}"
  register: cadvisor_container_ip

- debug: var=cadvisor_container_ip
```

I got:
 template error while templating string: unexpected '.'

I am probably doing something wrong. but I can see the stdout of the command is correct in debug mode.

Thanks,
Alberto.

**Comment by marconius** [comment]
For what it's worth, I am also getting this error when running

``` yaml
- name: register superweb container ip address
  command: "docker inspect --format {% raw %}'{{ .NetworkSettings.IPAddress }}'{% endraw %} superweb"
  register: superweb_ip

- debug: msg="IP is {{ superweb_ip.stdout }}"
```

playing around with ipdb, I found the error occurs during templating of the `debug` task.

**Comment by bcoca** [comment]
~~.NetworkSettings.IPAddress is not a valid variable name, you cannot start by a ., valid characters are letters, numbers and _~~
nvmd, that is what is happening but not what you are trying to do, {% raw %} does not seem to be working

**Comment by bcoca** [comment]
I tested with both current devel and 1.9.2 and it seems to work fine in both

**Comment by bcoca** [comment]
can you try either/both versions?

**Comment by marconius** [comment]
I tried both and fails on both. What is weird to me is that the template string that causes the trouble is not even present in the `debug` task in which the error is caught.

2.0.0

``` shell
PLAY: ***************************************************************************

TASK [setup] ********************************************************************
ok: [localhost]

TASK [website : website : register superweb container ip address] ***
changed: [localhost]
 [ERROR]: ERROR! template error while templating string: unexpected '.'
```

and 1.9.2

``` shell
PLAY [all] ******************************************************************** 

GATHERING FACTS *************************************************************** 
ok: [localhost]

TASK: [website | register superweb container ip address] ************* 
changed: [localhost]

TASK: [website | debug msg="IP is {{ superweb_ip.stdout }}"] ********* 
fatal: [localhost] => Failed to template msg="IP is {{ superweb_ip.stdout }}": Failed to template docker inspect --format '{{ .NetworkSettings.IPAddress }}' superweb: template error while templating string: unexpected '.'

FATAL: all hosts have already failed -- aborting

PLAY RECAP ******************************************************************** 
           to retry, use: --limit @/home/marco/all.retry

localhost                  : ok=2    changed=1    unreachable=1    failed=0
```

**Comment by thoraxe** [comment]
So this seems to be related to:
https://github.com/ansible/ansible/issues/10156

I am also seeing this myself. Similar kind of thing:

``` yaml
  - command: oc get service docker-registry -t "{% raw %}{{ .spec.portalIP }}:{{ with index .spec.ports 0 }}{{ .port }}{{ end }}{% endraw %}"
    register: registry_uri_out

  - set_fact:
      registry_uri: "{{ registry_uri_out.stdout }}"
```

And I get the same kind of error:

```
fatal: [ec2-54-86-64-109.compute-1.amazonaws.com] => Failed to template {{ registry_uri_out.stdout }}: Failed to template oc get service docker-registry -t "{{ .spec.portalIP }}:{{ with index .spec.ports 0 }}{{ 
.port }}{{ end }}": template error while templating string: unexpected '.'

FATAL: all hosts have already failed -- aborting
```

This post seems to indicate weird escaping can help:
http://adeslade.co.uk/post/docker-inspect-inside-an-ansible-task-with-a-go-template

I'm testing that right now.

Any other ideas here?

**Comment by thoraxe** [comment]
No dice on the weird escaping. I can't seem to make that work at all (in the command).

**Comment by thoraxe** [comment]
@enxebre would you mind changing the title of this issue? I think it's wider-reaching than just when using `{%raw%}`

What do you think about the following:

"Accessing registered variables created from commands with braces results in template processing errors"

**Comment by sivel** [comment]
I've been unable to duplicate this either.  A few questions:
1. What is the output of ansible-playbook with `-v`?
2. What is `/bin/sh` on the target server?
3. What happens when you run your command manually when using the `/bin/sh` shell?
4. If you specify a different `executable` for your `command` task such as `/bin/bash` does the same problem happen?

**Comment by thoraxe** [comment]
Some more info (-vvvv output)

```
TASK: [command oc get service docker-registry -t "{{ .spec.portalIP }}:{{ with index .spec.ports 0 }}{{ .port }}{{ end }}"] *** 
<ec2-54-86-64-109.compute-1.amazonaws.com> ESTABLISH CONNECTION FOR USER: openshift
<ec2-54-86-64-109.compute-1.amazonaws.com> REMOTE_MODULE command oc get service docker-registry -t "{{ .spec.portalIP }}:{{ with index .spec.ports 0 }}{{ .port }}{{ end }}"
<ec2-54-86-64-109.compute-1.amazonaws.com> EXEC ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=900s -o ControlPath="/home/thoraxe/.ansible/cp/%h-%r" -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=openshift -o ConnectTimeout=10 ec2-54-86-64-109.compute-1.amazonaws.com /bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1444182284.86-239491309106966 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1444182284.86-239491309106966 && echo $HOME/.ansible/tmp/ansible-tmp-1444182284.86-239491309106966'
<ec2-54-86-64-109.compute-1.amazonaws.com> PUT /tmp/tmp0YMLQi TO /home/openshift/.ansible/tmp/ansible-tmp-1444182284.86-239491309106966/command
<ec2-54-86-64-109.compute-1.amazonaws.com> EXEC ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=900s -o ControlPath="/home/thoraxe/.ansible/cp/%h-%r" -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=openshift -o ConnectTimeout=10 ec2-54-86-64-109.compute-1.amazonaws.com /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=yvelnhlvuwdnjsilqirckxwrirsdykdj] password: " -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-yvelnhlvuwdnjsilqirckxwrirsdykdj; LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/bin/python /home/openshift/.ansible/tmp/ansible-tmp-1444182284.86-239491309106966/command; rm -rf /home/openshift/.ansible/tmp/ansible-tmp-1444182284.86-239491309106966/ >/dev/null 2>&1'"'"''
changed: [ec2-54-86-64-109.compute-1.amazonaws.com] => {"changed": true, "cmd": ["oc", "get", "service", "docker-registry", "-t", "{# .spec.portalIP #}:{# with index .spec.ports 0 #}{# .port #}{# end #}"], "delta": "0:00:00.227642", "end": "2015-10-06 21:44:46.215995", "rc": 0, "start": "2015-10-06 21:44:45.988353", "stderr": "", "stdout": "172.30.128.33:5000", "warnings": []}

TASK: [debug ] **************************************************************** 
<ec2-54-86-64-109.compute-1.amazonaws.com> ESTABLISH CONNECTION FOR USER: openshift
fatal: [ec2-54-86-64-109.compute-1.amazonaws.com] => Failed to template {{registry_uri_out.stdout_lines}}: Failed to template oc get service docker-registry -t "{{ .spec.portalIP }}:{{ with index .spec.ports 0 }}{{ .port }}{{ end }}": template error while templating string: unexpected '.'

FATAL: all hosts have already failed -- aborting
```

The value of stdout appears to be  "172.30.128.33:5000"

The command is registered to the variable "registry_uri_out", so I would assume that "registry_uri_out.stdout" would only contain the string... but when trying to use registry_uri_out.stdout in the subsequent task, Ansible blows up with the error. 

Note that the pasted content above is where I just try to debug registry_uri_out.stdout as opposed to actually using it in "set_fact" but the result is exactly the same -- a complaint about "failed to template".
1. 

```
[openshift@ip-172-18-1-193 ~]$ ls -l `which sh`
lrwxrwxrwx. 1 root root 4 Feb 25  2015 /usr/bin/sh -> bash
[openshift@ip-172-18-1-193 ~]$ bash --version
GNU bash, version 4.2.46(1)-release (x86_64-redhat-linux-gnu)
Copyright (C) 2011 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```
1. 

```
oc get service docker-registry -t "{{ .spec.portalIP }}:{{ with index .spec.ports 0 }}{{ .port }}{{ end }}"
172.30.128.33:5000
```
1. `sh` is symlinked to `bash` so I'm not sure if that would change anything?

**Comment by michaljrk** [comment]
Hello,

Did anybody find a workaround for this issue?

**Comment by michaljrk** [comment]
Any actions planned to fix the issue?

**Comment by thoraxe** [comment]
I've just hit this again in a completely different way:

https://gist.github.com/thoraxe/046a0abd91f0a34d6641ceb95bd3ab5d

This time I get the error about 20 tasks later after using { %raw }.

**Comment by michaljrk** [comment]
I have eventually managed to workaround with { %raw }, like this:

> docker inspect -f \"{% raw %}{{ .Config.Cmd }}{% endraw %}\" mongo

but still, it took ages to get there and would be nice to have the bug fixed.

**Comment by thoraxe** [comment]
So, I noticed that even if you have {%raw%} randomly commented out somewhere, that Ansible still explodes somewhere arbitrarily.

So it appears that there is a bug, in general, with the use of {%raw%}...

**Comment by ansibot** [comment]
@enxebre ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@enxebre closure has been requested for this issue or it has timed out waiting for your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by pixelrebel** [comment]
I have the same issue in 2.1.2.0 on Sierra.
Task:

```
- name: This should work
  debug: var="{{ playbook_dir }}"
```

Error:

```
TASK [test-role : This should work] **************************************************
fatal: [server]: FAILED! => {"failed": true, "msg": "template error while templating string: unexpected '/'. String: {{/Users/pixelrebel/dev/plays}}"}
```

I wouldn't close this out just yet @ansibot

**Comment by wilddog64** [comment]
Further more, something like this: RUX<MiYr+8Em will also trigger the problem. The string is extracted from mysqld.log when performing mysql reset root password ...

**Comment by thoraxe** [comment]
Why is this closed? It's still a problem?

**Comment by pixelrebel** [comment]
@thoraxe Yes this is still an issue in 2.2.0.

**Comment by pixelrebel** [comment]
@thoraxe My example above may be syntactically incorrect.  

```
- name: This should work
  debug: var=playbook_dir 
```
works, but I can't recall what caused me to try this example.    I hit this this bug on another task and then created the debug example for reference.  I haven't come across this bug since.  Perhaps I implemented it incorrectly.

**Comment by tumbl3w33d** [comment]
This error happened in my playbook coming from nowhere. I didn't touch it, it was working for months. I just sometimes pull from devel [currently ansible 2.3.0 (devel 8495b6323a)], but even going back some revisions I couldn't make it work again.

I had
```
docker inspect -f {% raw %}'{{ .NetworkSettings.Networks.bridge.IPAddress }}'{% endraw %} {{ container_name }}
```
and had to change it to
```
docker inspect -f \{\{.NetworkSettings.Networks.bridge.IPAddress\}\} {{ container_name }}
```
to make it work again.

**Comment by ravishivt** [comment]
Update:  I am hitting https://github.com/ansible/ansible/issues/20568.  Downgrading to 2.1.4 fixed the issue.

~~This issue should be reopened.~~

I think the common pattern with everyone reporting the issue, including me, is the dot "." as the first character inside the curly braces, e.g. {{ .ABC }}.   That seems to be the triggering condition that can't be escaped.  Also, ansible doesn't error on the task that contains the dotted variable, only on subsequent commands that try to use its registered variable.

For me, the tasks are:

```
- name: Get the Heketi Server Endpoint
  shell: kubectl get svc/heketi --template "http://{{ '{{' }}.spec.clusterIP{{ '}}' }}:{{ '{{' }}(index .spec.ports 0).port{{ '}}' }}"
  # shell: kubectl get svc/heketi --template {%raw%}'http://{{.spec.clusterIP}}:{{(index .spec.ports 0).port}}'{%endraw%}
  register: heketi_endpoint

- debug: var=heketi_endpoint.stdout
```

With output of:

```
TASK [gk-post : Get the Heketi Server Endpoint] ********************************
changed: [ravi-kube196]

TASK [gk-post : debug] *********************************************************
fatal: [ravi-kube196]: FAILED! => {"failed": true, "msg": "{u'cmd': u'kubectl get svc/heketi --template \"http://{{.spec.clusterIP}}:{{(index .spec.ports 0).port}}\"', u'end': u'2017-01-24 19:39:06.448667', u'stdout': u'http://10.102.192.231:8080', u'changed': True, u'start': u'2017-01-24 19:39:06.363425', u'delta': u'0:00:00.085242', u'stderr': u'', u'rc': 0, 'stdout_lines': [u'http://10.102.192.231:8080'], u'warnings': []}: template error while templating string: unexpected '.'. String: kubectl get svc/heketi --template \"http://{{.spec.clusterIP}}:{{(index .spec.ports 0).port}}\""}
```

Anyone figure out a workaround?

**Comment by Chabane** [comment]
I have the same issue 

```
 - name: set up a port-forward
    shell : >
          POD=$(kubectl get pods --namespace kube-system -l k8s-app=kube-registry -o template --template '{{range .items}}{{.metadata.name}} {{.status.phase}}{{"\n"}}{{end}}' | grep Running | head -1 | cut -f1 -d' ')
```

```
fatal: [nirby-manager1]: FAILED! => {"failed": true, "msg": "template error while templating string: unexpected '.'. String: 'POD=$(kubectl get pods --namespace kube-system -l k8s-app=kube-registry -o template --template '{{range .items}}{{.metadata.name}} {{.status.phase}}{{\"\\\\n\"}}{{end}}' | grep Running | head -1 | cut -f1 -d' ')'"}
```

**Comment by jakubigla** [comment]
Ansible became a **** joke. That's another issue I'm struggling today because of version updates.

**Comment by jakubigla** [comment]
My workaround:

```
- name: Register mongo service IP
  shell: "docker inspect mongo"
  register: inspect_output

- set_fact:
    mongoIp: "{{ item.NetworkSettings.Networks[docker_network_name].IPAddress }}"
  with_items: "{{ inspect_output.stdout|from_json }}"
```

Where `docker_network_name` is a variable with my custom network name.

**Comment by dmsimard** [comment]
Just cross-referencing, this seems to be the same issue as https://github.com/ansible/ansible/issues/20400 ?

**Comment by fat0** [comment]
Thanks for the downgrade suggestion @ravishivt. Ansible in later versions is still broken when the value in the var being parsed has special characters like a hyphen (`-`). As a testing example, folx can try the following test which uses Ansible to get the latest AWS EC2 `ami_id` of your liking: https://recursive.cloud/blog/latest-amazon-ami/. For testing purposes, if you replace `ami_id` with a key that has a value without a `-`, eg `owner_id`, then the parsing works fine (i.e replace the last line with `var: "{{ amis.results[0].owner_id }}"`)

**Comment by dagwieers** [comment]
@fat0 and others, this repository is no longer active. If you have something to share and actually want people to see it and help with it, please open a new ticket at: https://github.com/ansible/ansible/issues/new
